### DISCUSSION TITLE
TreeView duplicate entry bug

### DISCUSSION THREAD

**Discussion by MajesticBevans** [AvaloniaUI/Avalonia#17274] [Q&A]
### Describe the bug

I have a treeview that is bound to an ItemsSource (a static, readonly ObservableCollection of a custom class) containing one item. That item has children, and its children may also have children of their own. Occasionally, when I add this single item to the collection bound to the TreeView, the top-level parent object is duplicated in the TreeView, along with all its children and grandchildren. This erroneous duplicate object can be expanded and interacted with as a normal object, but affects the state of the original object as well. (see video for what I mean). 

I have made certain that I am not adding the item to the collection twice, and Avalonia is clearly recognising them as the same object, despite duplicating them in its displayed entries, as they share the same functionality.

This duplicate object will remain in the UI, however, even when the actual original object is removed from the collection.

Video of bug can be found here: https://viperapplied-my.sharepoint.com/:v:/g/personal/ben_viper_as/EfIMp-8Sh3lBm6TfIJY1vTEBWKz3QJr_BUMGz105wNDOlA?e=3kgi6B

### To Reproduce

I cannot share the details of my code as it is for commercial use, but here is my TreeView definition: 

```
<TreeView
    AutoScrollToSelectedItem="True"
    ItemsSource="{Binding OpenDirectories}"
    SelectedItem="{Binding SelectedObject}"
    SelectionMode="AlwaysSelected, Single">
    <TreeView.DataTemplates>
        <!--  Tree Data Template for Directory  -->
        <TreeDataTemplate DataType="r:SelectableDirectory" ItemsSource="{Binding Children}">
            <tv:DirectoryRenderTreeView DataContext="{Binding}" />
        </TreeDataTemplate>

        <!--  Tree Data Template for Volume Series  -->
        <TreeDataTemplate DataType="r:VolumeSeries" ItemsSource="{Binding Children}">
            <tv:VolumeSeriesTreeView DataContext="{Binding}" />
        </TreeDataTemplate>

        <!--  Tree Data Template for Surface Series  -->
        <TreeDataTemplate DataType="r:SurfaceSeries" ItemsSource="{Binding Children}">
            <tv:SurfaceSeriesTreeView DataContext="{Binding}" />
        </TreeDataTemplate>

        <!--  Tree Data Template for Obstacle Series  -->
        <TreeDataTemplate DataType="r:MultiObstacleSeries" ItemsSource="{Binding Children}">
            <tv:MultiObstacleSeriesTreeView DataContext="{Binding}" />
        </TreeDataTemplate>
    </TreeView.DataTemplates>
</TreeView>
```

The ItemsSource collection definition:

`public static ObservableCollection<SelectableDirectory> OpenDirectories => _openDirectories;`

and the SelectedItem definition:

```
 public IRenderable? SelectedObject
    {
        get => _selectedObject;
        set => SetProperty(ref _selectedObject, value); 
    }
```

The issue is that I cannot reliably reproduce this behaviour, it just happens sometimes. Even when using exactly the same objects as the itemsource I can't reliably reproduce it.

Sorry for the limited info. Has anyone encountered this behaviour before?

### Expected behavior

I expect the treeview item to not be duplicated in the UI. Everything else, is fine. Even the functionality remains intact, it seems to simply be a UI bug.

### Avalonia version

11.1.3

### OS

Windows

**Comment by timunie** [comment]
With out code to reproduce, we cannot investigate anything. If you need support for a commercial App, I suggest to to consider a support agreement.

**Reply by MajesticBevans** [reply]
I understand. I'll try and find a way to reliably reproduce it

**Reply by dr3adl0rd** [reply]
I just got the same problem today, and I figured out the reason. 
When a View is instantiated and the code to populate the property that is bound to the ItemSource of a TreeView is executed asynchronously, i.e. via Task.Run for example, then it will show duplicate contents of the collection bound via ItemSource if the async method is also using await.

i.e.
```
public partial class MyView : ObservableObject
{
...
public ObservableCollection<TreeNode> Nodes {get; set;} = [];

public MyView() {
    Task.Run(LoadData);
}

private async Task LoadData()
{
    Nodes.Clear();
    await DoSomethingAsync();
    Nodes.Add(new TreeNode("text 1")); // Tree will contain text 1 twice!!!!
}
```

UPDATE 1:
Actually duplication happens whenever the Bound property, e.g. Nodes is updated inside a async call.

**Reply by timunie** [reply]
That's by design. Collection updates should come from the UIThread. You can use the Dispatcher inside async Task if needed
